lib/identity_cache/query_api.rb

```diff
@@ -154,6 +148,25 @@ def cache_fetch_includes
       end
     end
 
+    included do |base|
+      # Make sure there is at least once after_commit callback so that _run_commit_callbacks
+      # is called, which is overridden to do an early after_commit callback
+      base.after_commit {}
+    end
+
+    # Override the method that is used to call after_commit callbacks so that we can
+    # expire the caches before other after_commit callbacks. This way we can avoid stale
+    # cache reads that happen from the ordering of callbacks. For example, if an after_commit
+    # callback enqueues a background job, then we don't want it to be possible for the
+    # background job to run and load data from the cache before it is invalidated.
+    def _run_commit_callbacks
+      if destroyed? || transaction_changed_attributes.present?
+        expire_cache
+        expire_parent_caches
+      end
```

Comment:
nit: this could be factored into a function `identity_cache_expire_caches` only reason I'm suggesting that is that would make it more clear in a stack trace what is going on. Take it or leave it not a strong feeling here.